### ISSUE TITLE
Deprecate global `paths` setting and reduce `which` calls

### ISSUE THREAD

**Issue by kaste** [SublimeLinter/SublimeLinter#1199]
Usually global binaries are already on PATH. It is not necessary to call `which` in most cases bc it makes no difference to call `path/to/node` for global binaries. You can just call `node` and it will work. If the binary is not there, the OS will report an error, which we can parse and show the user. Our `which` call is also naive, users could execute remote binaries, but our own `which` prevents this.

People use `paths` e.g. to circumvent the OSX startup issue. But you cannot have multiple `node`'s here. 

For SL4, users can enhance the PATH using the `env` setting.

**Comment by kaste** [comment]
Notable: `paths` is only used for our `which` calls. The linter has its own env and does not inherit this augmented path.

**Comment by ChisholmKyle** [comment]
I have an issue related to this. I'm working on a linter that has two steps: 

1. Run cppcheck on a source file (temporary file from SublimeLinter)
    ```sh
    cppcheck --dump tempfile.c
    ```
2. This produces tempfile.c.dump and I need to send that to an addon
    ```sh
    python addon.py tempfile.c.dump
    ```

My problem is that in the `linter.py` I cannot simply do
```python
    cmd = 'cppcheck --dump @ && python addon.py @.dump'
```

because `&&` is interpreted as an input to `cppcheck`. So I tried to create an executable shell script called `cppcheck-addon` that takes the file as input and runs the command, so that I set 
```python
    cmd = 'cppcheck-addon @'
```

But now the PATH is messed up and I get the error from the script `cppcheck-addon`: `cppcheck: command not found`. Setting PATH as an environment variable works, but it seems unreasonable to have to do this when it is already properly set.

**Comment by kaste** [comment]
> when it is already properly set.

What do you mean by that?

**Comment by ChisholmKyle** [comment]
I mean in `~/.profile` or similar. I'm using Macports and zsh as default, so the modified PATH is set in `.zshrc`. It turns out I just had to wait a few moments and the linter reloads without the error. I think the linter runs before the PATH is set on startup. I'm not really sure how environemtn variables are loaded with Sublime Text and Package Control.

**Comment by kaste** [comment]
Okay, that just means you have relatively slow login shell.

**Comment by FichteFoll** [comment]
TL;DR: Because of how applications are launched on macOS, they don't inherit the environment of the login shell in many cases and ST has to retrieve them from a login shell session manually. However, this happens asynchronously and plugins may run before this happens, leading to this problem.

**Comment by braver** [comment]
Hid some comments because loading the env has changed completely in ST 3.1.

**Comment by kaste** [comment]
Notable: For Windows we use the which calls to get proper file extensions. Only '.exe' can be omitted. '.bat', '.cmd' *must* be present (See $PATHEXT env var.) or it will fail.

**Comment by kaste** [comment]
Closing, this is not actionable and the (my) initial description and title are probably wrong. At least they confuse me. We start a new issue about our old-school `paths` at some time.
